docs/proposals/20240807-in-place-updates.md

```diff
@@ -0,0 +1,927 @@
+
+On each iteration, the controller will substract from the set of desired changes the subset returned by the updater. 
+
+If this set is reduced to zero, then CAPI will determine that the update can be performed using the external strategy. CAPI will define the update plan as a list of sequential external updaters in a particular order and proceed to execute it. The update plan will be stored in the Machine object as an array of strings (the names of the selected external updaters).
+
+If after iterating over all external updaters the remaining set still contains uncovered changes, CAPI will determine the desired state cannot be reached through external updaters. If a fallback rolling update strategy has been configured (this is optional), CAPI will replace the machines. If no fallback strategy is configured, we will surface the issue in the resource status. Machines will remain unchanged and the desired state won't be reached unless remediated by the user. Depending on the scenario, users can: ammend the desired state to something that the registered updaters can cover, register additional updaters capable of handling the desired changes or simply enable the fallback strategy.
+
+### MachineDeployment updates
+
+```mermaid
+sequenceDiagram
+participant Operator
+
+box Management Cluster
+    participant apiserver as kube-api server
+    participant capi as MD controller
+    participant msc as MachineSet Controller
+    participant mach as Machine Controller
+    participant hook as External updater
+end
+
+Operator->>apiserver: make changes to MD <br>(ExternalUpdate strategy)
+apiserver->>capi: Notify changes
+apiserver->>Operator: OK
+capi->>capi: Define Update Plan
+capi->>apiserver: Create new MachineSet
+loop For all machines
+    capi->>apiserver: Add Update Plan to Machine<br> and move to new Machine Set
+    apiserver->>msc: Notify changes
+    msc->>apiserver: Update Machine Spec
+    apiserver->>mach: Notify changes
+    loop For all updaters in plan
+        mach->>hook: Run updater
+    end
+    mach->>apiserver: Mark Machine as updated
+end
+```
+
+The MachineDeployment controller updates machines in place in a very similar way to rolling updates: by creating a new MachineSet and moving the machines from the old MS to the new one. We want to stress that the Machine objects won't be deleted and recreated like in the current rolling strategy. The MachineDeployment will just update the OwnerRefs, effectively moving the existing Machine object from one MS to another. The number of machines moved at once might be made configurable on the MachineDeployment in the same way `maxSurge` and `maxUnavailable` control this for rolling updates.
```

Comment:
nit: in addition to the OwnerRef, a label on the Machine will need to be updated to match the selector in the new MachineSet.